pika/adapters/select_connection.py

```diff
@@ -217,120 +195,202 @@ def add_timeout(self, deadline, callback_method):
+    def process_timeouts(self):
+        """Process the self._timeouts event stack"""
+        
+        now = time.time()
+        to_run = filter(lambda t: t['deadline'] <= now,
+                        self._timeouts.values())
+
+        # Run the timeouts in order of deadlines. Although this shouldn't
+        # be strictly necessary it preserves old behaviour when timeouts
+        # were only run periodically.
+        for t in sorted(to_run, key=itemgetter('deadline')):
+            t['callback']()
+            del self._timeouts[hash(frozenset(t.items()))]
+            self._next_timeout = None
+
+    def add_handler(self, fileno, handler, events):
+        """Add a new fileno to the set to be monitored
+
+       :param int fileno: The file descriptor
+       :param method handler: What is called when an event happens
+       :param int events: The event mask
+
+       """
+        self._fd_handlers[fileno] = handler
+        self.update_handler(fileno, events)
+
+    def update_handler(self, fileno, events):
+        """Set the events to the current events
+
+        :param int fileno: The file descriptor
+        :param int events: The event mask
+
+        """
+
+        for ev in (READ, WRITE, ERROR):
+            if events & ev:
+                self._fd_events[ev].add(fileno)
+            else:
+                self._fd_events[ev].discard(fileno)
+    
+
+    def remove_handler(self, fileno):
+        """Remove a file descriptor from the set
+        
+        :param int fileno: The file descriptor
 
         """
-        # Build our values to pass into select
-        input_fileno, output_fileno, error_fileno = [], [], []
+        try:
+            del self._processing_fd_event_map[fileno]
+        except KeyError:
+            pass
+
+        self.update_handler(fileno, 0)
+        del self._fd_handlers[fileno]
+
+    def start(self):
+        """Start the main poller loop. It will loop here until self._stopping"""
+        
+        LOGGER.debug('Starting IOLoop')
+        self._stopping = False
+        
+        while not self._stopping:
+            self.poll()
+            self.process_timeouts()
 
-        if self.events & READ:
-            input_fileno = [self.fileno]
-        if self.events & WRITE:
-            output_fileno = [self.fileno]
-        if self.events & ERROR:
-            error_fileno = [self.fileno]
+    def stop(self):
+        """Exit from the ioloop. """
 
-        # Wait on select to let us know what's up
+        LOGGER.debug('Stopping IOLoop')
+        self._stopping = True
+        
         try:
-            read, write, error = select.select(input_fileno, output_fileno,
-                                               error_fileno,
-                                               SelectPoller.TIMEOUT)
-        except select.error as error:
-            return self._handler(self.fileno, ERROR, error)
+            # Send byte to interrupt the poll loop
+            self._w_interrupt.send('X')
```

Comment:
Yes, again a fairly theoretical one, but we're still catching it. If we do get EWOULDBLOCK then it means we've sent an awful lot of unread interrupt bytes so there's not much we can do except log it.

WRT send/write. The ioloop works in terms of filedescriptors so the code is a lot cleaner simply doing read/write on those. This could be changed to os.write(self._w_interrupt.fileno(), 'X') but I don't think it adds anything. Happy to change if you feel strongly otherwise.